include/deal.II/multigrid/mg_transfer_matrix_free.h

```diff
@@ -417,16 +490,26 @@ copy_to_mg (const DoFHandler<dim,spacedim> &mg_dof,
   // FIXME: this a quite ugly as we need a temporary object:
   MGLevelObject<LinearAlgebra::distributed::Vector<Number>> dst_non_block(min_level, max_level);
 
-  for (unsigned int b = 0; b < n_blocks; ++b)
-    {
-      for (unsigned int l = min_level; l <= max_level; ++l)
-        dst_non_block[l].reinit(dst[l].block(b));
+  if (same_for_all)
```

Comment:
the branches here are exactly the same with a difference of `matrix_free_transfer_vector[0]` vs `matrix_free_transfer_vector[b]`. Maybe to avoid code duplication add a  lambda function which returns `0` if `same_for_all` or `b` otherwise? Same for copy_from_mg. Or hide it into internal function which returns a transfer object based on `b` and `same_for_all`.